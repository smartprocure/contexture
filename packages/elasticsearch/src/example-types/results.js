let F = require('futil')
let _ = require('lodash/fp')
let {
  highlightResults,
  arrayToHighlightsFieldMap,
} = require('../utils/highlighting')
let { getField } = require('../utils/fields')

module.exports = {
  validContext: () => true,
  result(node, search, schema) {
    let page = (node.page || 1) - 1
    let pageSize = node.pageSize || 10
    let startRecord = page * pageSize
    let sortField = node.sortField ? getField(schema, node.sortField) : '_score'

    let searchObj = {
      from: startRecord,
      size: pageSize,
      sort: { [sortField]: node.sortDir || 'desc' },
      explain: node.explain,
      // Without this, ES7+ stops counting at 10k instead of returning the actual count
      track_total_hits: true,
    }

    if (node.include || node.exclude)
      searchObj._source = F.compactObject({
        includes: node.include,
        excludes: node.exclude,
      })

    // Global schema highlight configuration
    let schemaHighlight =
      node.highlight !== false && schema.elasticsearch.highlight
    // Specific search highlight override
    let searchHighlight = _.isPlainObject(node.highlight) ? node.highlight : {}
    let resultColumns = node.include

    // Highlighting starts with defaults in the schema first
    if (schemaHighlight) {
      let showOtherMatches = _.getOr(false, 'showOtherMatches', node)
      let schemaInline = _.getOr([], 'inline', schemaHighlight)
      let schemaInlineAliases = _.flow(
        _.getOr({}, 'inlineAliases'),
        _.entries,
        _.filter(([k]) => _.includes(k, node.include)),
        _.flatten
      )(schemaHighlight)

      // Concat the search specific override fields with the schema `inline` so we have them as targets for highlight replacement
      schemaHighlight = _.set(
        'inline',
        _.concat(schemaInline, _.keys(searchHighlight.fields)),
        schemaHighlight
      )
      // Convert the highlight fields from array to an object map
      let fields = _.flow(
        _.pick(['inline', 'additionalFields']), // Get the highlight fields we will be working with
        _.values,
        _.flatten,
        _.concat(schemaInlineAliases), // Include the provided field aliases if any
        _.uniq,
        arrayToHighlightsFieldMap, // Convert the array to object map so we can simply _.pick again
        filtered =>
          showOtherMatches
            ? // Highlight on all fields specified in the initial _.pick above.
              filtered
            : // Only highlight on the fields listed in the node include section and their aliases (if any)
              _.pick(_.concat(node.include, schemaInlineAliases), filtered)
      )(schemaHighlight)

      // Setup the DEFAULT highlight config object with the calculated fields above
      // and merge with the search specific config
      searchObj.highlight = _.merge(
        {
          // The default schema highlighting settings w/o the fields
          pre_tags: ['<b class="search-highlight">'],
          post_tags: ['</b>'],
          require_field_match: false,
          number_of_fragments: 0,
          fields,
        },
        searchHighlight
      )

      // Make sure the search specific overrides are part of the node include.
      // This way they do not have to be added manually. All that is needed is the highlight config
      resultColumns = _.flow(
        _.concat(_.keys(searchHighlight.fields)),
        _.uniq,
        _.compact
      )(node.include)

      // Make sure search returns the resultColumns we want by setting the _.source.includes
      F.setOn('_source.includes', resultColumns, searchObj)
    }

    return search(searchObj).then(results => ({
      scrollId: results._scroll_id,
      // ES 7+ is total.value, ES 6- is hits.total
      totalRecords: F.getOrReturn('value', results.hits.total),
      startRecord: startRecord + 1,
      endRecord: startRecord + results.hits.hits.length,
      results: _.map(hit => {
        let additionalFields
        if (schemaHighlight) {
          let highlightObject = highlightResults(
            schemaHighlight, // The highlight configuration
            hit, // The ES result
            schema.elasticsearch.nestedPath,
            resultColumns // The columns to return
          )
          additionalFields = highlightObject.additionalFields
        }

        // TODO - If nested path, iterate properties on nested path, filtering out nested path results unless mainHighlighted or relevant nested fields have "<b></b>" tags in them
        return {
          additionalFields: schemaHighlight ? additionalFields : [],
          ...hit,
        }
      }, results.hits.hits),
    }))
  },
}
