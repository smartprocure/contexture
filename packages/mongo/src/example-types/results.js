let F = require('futil')
let _ = require('lodash/fp')

let convertPopulate = getSchema =>
  _.flow(
    F.mapIndexed((x, as) => {
      let { unwind, schema } = x
      let targetSchema = getSchema(schema) //|| toSingular(as), //<-- needs compromise-fp
      if (!targetSchema)
        throw Error(`Couldn't find schema configuration for ${schema}`)
      if (!targetSchema.mongo)
        throw Error(
          'Populating a non mongo provider schema on a mongo provider schema is not supported'
        )
      let targetCollection = _.get('mongo.collection', targetSchema)
      if (!targetCollection)
        throw Error(
          `The ${targetCollection} schema has a mongo configuration, but doesn't have a 'collection' property`
        )

      let $unwind = unwind
        ? [
            {
              $unwind: {
                path: `$${as}`,
                preserveNullAndEmptyArrays: true,
              },
            },
          ]
        : []
      let $lookup = [
        {
          $lookup: {
            as,
            from: targetCollection,
            localField: x.localField, // || '_id',
            foreignField: x.foreignField, // || node.schema, <-- needs schema lookup
          },
        },
      ]
      return [...$lookup, ...$unwind]
    }),
    _.flatten
  )

let getStartRecord = ({ page, pageSize }) => {
  page = page < 1 ? 0 : page - 1
  return page * pageSize
}

let parentPath = path => path.replace(/(\.[^.]+)$/, '')

let isParentPathProjected = include => path =>
  _.some(_.eq(parentPath(path)), _.pull(path, include))

let projectFromInclude = include =>
  _.flow(
    _.remove(isParentPathProjected(include)),
    _.countBy(_.identity)
  )(include)

let getResultsQuery = (node, getSchema, startRecord) => {
  let { pageSize, sortField, sortDir, populate, include } = node

  // $sort, $skip, $limit
  let $sort = {
    $sort: {
      [sortField]: sortDir === 'asc' ? 1 : -1,
    },
  }
  let $limit = { $limit: pageSize }
  let sortSkipLimit = _.compact([
    sortField && $sort,
    { $skip: startRecord },
    pageSize > 0 && $limit,
  ])
  // If sort field is a join field move $sort, $skip, and $limit to after $lookup.
  // Otherwise, place those stages first to take advantage of any indexes on that field.
  let sortOnJoinField = _.some(
    x => _.startsWith(`${x}.`, sortField) || sortField === x,
    _.keys(populate)
  )
  // $project
  let $project = _.isEmpty(include)
    ? []
    : [{ $project: projectFromInclude(include) }]

  return [
    ...(!sortOnJoinField ? sortSkipLimit : []),
    ...convertPopulate(getSchema)(populate),
    ...(sortOnJoinField ? sortSkipLimit : []),
    ...$project,
  ]
}

let defaults = _.defaults({
  page: 1,
  pageSize: 10,
  sortDir: 'desc',
  include: [],
})

let result = async (node, search, schema, { getSchema }) => {
  node = defaults(node)
  let startRecord = getStartRecord(node)
  let resultsQuery = getResultsQuery(node, getSchema, startRecord)
  let countQuery = [{ $group: { _id: null, count: { $sum: 1 } } }]

  let [results, count] = await Promise.all([
    search(resultsQuery),
    !node.skipCount && search(countQuery),
  ])

  return {
    response: {
      totalRecords: _.get('0.count', count),
      startRecord: startRecord + 1,
      endRecord: startRecord + results.length,
      results,
    },
  }
}

// NOTE: pageSize of 0 will return all records
module.exports = {
  getStartRecord,
  getResultsQuery,
  defaults,
  projectFromInclude,
  convertPopulate,
  // API
  result,
}
