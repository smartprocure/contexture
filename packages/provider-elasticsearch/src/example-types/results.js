let F = require('futil-js')
let _ = require('lodash/fp')
let { highlightResults, arrayToHighlightsFieldMap } = require('../highlighting')
let { getField } = require('../fields')

module.exports = {
  result(context, search, schema) {
    let page = (context.page || 1) - 1
    let pageSize = context.pageSize || 10
    let startRecord = page * pageSize
    let sortField = context.sortField
      ? // default to word (aka '') for backwards compatibility
        getField(schema, context.sortField, context.sortMode || 'word')
      : '_score'
    let sortDir = context.sortDir || 'desc'
    let result = {
      from: startRecord,
      size: pageSize,
      sort: {
        [sortField]: sortDir,
      },
      explain: context.explain,
    }

    if (context.forceExclude && _.isArray(schema.forceExclude))
      context.exclude = _.union(schema.forceExclude, context.exclude)

    if (context.include || context.exclude) result._source = {}
    if (context.include) result._source.includes = context.include
    if (context.exclude) result._source.excludes = context.exclude

    let highlight =
      _.getOr(true, 'highlight', context) && schema.elasticsearch.highlight
    let inlineAliases = _.getOr({}, 'inlineAliases', highlight)
    let showOtherMatches = _.getOr(false, 'showOtherMatches', context)

    if (highlight) {
      // Convert the highlight fields from array to an object map
      let fields = _.flow(
        _.pick(['inline', 'additionalFields']),
        _.values,
        _.flatten,
        _.concat(_.values(inlineAliases)),
        _.uniq(),
        arrayToHighlightsFieldMap,
        filtered => showOtherMatches
          ? filtered
          : _.pick(_.intersection(context.include, _.keys(filtered)), filtered),
      )(highlight)

      F.extendOn(result, {
        highlight: {
          pre_tags: ['<b class="search-highlight">'],
          post_tags: ['</b>'],
          require_field_match: false,
          number_of_fragments: 0,
          fields
        },
      })
    }

    return search(result).then(results => ({
      scrollId: results._scroll_id,
      response: {
        totalRecords: results.hits.total,
        startRecord: startRecord + 1,
        endRecord: startRecord + results.hits.hits.length,
        results: _.map(hit => {
          let highlightObject
          let additionalFields
          if (highlight) {
            highlightObject = highlightResults(
              schema.elasticsearch.highlight,
              hit,
              schema.elasticsearch.nestedPath,
              context.include
            )
            additionalFields = highlightObject.additionalFields
          }

          // TODO - If nested path, iterate properties on nested path, filtering out nested path results unless mainHighlighted or relevant nested fields have b tags in them
          return _.extendAll([
            context.verbose || _.size(context.include) > 0 ? { hit } : {},
            {
              _id: hit._id,
              additionalFields: highlight ? additionalFields : [],
            },
            _.getOr(_.identity, 'elasticsearch.summaryView', schema)(hit),
          ])
        }, results.hits.hits),
      },
    }))
  },
}
