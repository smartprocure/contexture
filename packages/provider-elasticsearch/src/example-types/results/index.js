// https://stackoverflow.com/questions/70177601/does-elasticsearch-provide-highlighting-on-copy-to-field-in-their-newer-versio
// https://github.com/elastic/elasticsearch/issues/5172

import _ from 'lodash/fp.js'
import F from 'futil'
import {
  getHighlightFields,
  alignHighlightsWithSourceStructure,
} from './highlight.js'
import { getField } from '../../utils/fields.js'

const defaultHighlightConfig = {
  pre_tag: '<b class="search-highlight">',
  post_tag: '</b>',
}

export default {
  validContext: () => true,
  async result(node, search, schema) {
    const page = (node.page || 1) - 1
    const pageSize = node.pageSize || 10
    const startRecord = page * pageSize
    const sortField = node.sortField
      ? getField(schema, node.sortField)
      : '_score'

    const highlightConfig = _.defaults(defaultHighlightConfig, node.highlight)

    const body = F.omitBlank({
      from: startRecord,
      size: pageSize,
      sort: { [sortField]: node.sortDir || 'desc' },
      explain: node.explain,
      // Without this, ES7+ stops counting at 10k instead of returning the actual count
      track_total_hits: true,
      _source: F.omitBlank({ includes: node.include, excludes: node.exclude }),
      highlight: highlightConfig.enable && {
        pre_tags: [highlightConfig.pre_tag],
        post_tags: [highlightConfig.post_tag],
        number_of_fragments: 0,
        require_field_match: true,
        fields: getHighlightFields(schema, node._meta.relevantFilters),
      },
    })

    const response = await search(body)
    const results = response.hits.hits

    if (highlightConfig.enable) {
      // Not mutating source in the helper function leaves the door open
      // for a configuration flag to control inlining of highlighted
      // results in source
      const fn = alignHighlightsWithSourceStructure(schema, highlightConfig)
      for (const result of results) {
        for (const [field, val] of _.toPairs(fn(result))) {
          F.setOn(field, val, result._source)
        }
      }
    }

    return {
      scrollId: response._scroll_id,
      // ES 7+ is total.value, ES 6- is hits.total
      totalRecords: F.getOrReturn('value', response.hits.total),
      startRecord: startRecord + 1,
      endRecord: startRecord + results.length,
      results,
    }
  },
}
