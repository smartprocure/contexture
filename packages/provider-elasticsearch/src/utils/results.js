import { unsetOnTree } from './futil.js'

// This is the method that will rip off and collect all hoisted props in an array
export let hoistOnTree = unsetOnTree('__hoistProps')
