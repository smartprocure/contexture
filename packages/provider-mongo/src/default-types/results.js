let _ = require('lodash'),
  Promise = require('bluebird')

module.exports = {
  result: (context, search, schema) => {
    let page = (context.config.page || 1) - 1,
      pageSize = context.config.pageSize || 10,
      startRecord = page * pageSize,
      sortField = context.config.sortField || '_score',
      sortDir = context.config.sortDir || 'desc',
      sort = {
        [sortField]: sortDir === 'asc' ? 1 : -1
      }

    return Promise.all([
      search([{
        $sort: sort
      }, {
        $skip: startRecord
      }, {
        $limit: pageSize
      }]),
      search([{
        $group: {
          _id: null,
          count: {
            $sum: 1
          }
        }
      }])
    ]).spread((results, count) => ({
      // TODO - handle aggregate wrapped stuff, e.g. result.result or result.result[0] etc
      response: {
        totalRecords: _.get(count, '0.count'),
        startRecord: startRecord + 1,
        endRecord: startRecord + results.length,
        results: results
      }
    }))
  }
}
