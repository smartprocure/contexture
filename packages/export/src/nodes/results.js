import F from 'futil'
import _ from 'lodash/fp'
import { runWith, flattenProp } from '../utils'

// Will go away once results no longer wraps context in `response`
let resultField = (field, node) =>
  F.cascade([`response.${field}`, field], node.context)

export default async ({ service, tree, ...node }) => {
  let { page = 1, pageSize = 100 } = node
  let run = props =>
    runWith(
      service,
      tree,
      _.pickBy(_.negate(_.isNil), {
        key: 'results',
        type: 'results',
        ...node,
        ...props,
      })
    )
  let scrollId
  let totalRecords = resultField(
    'totalRecords',
    await run({ pageSize: 1, page: 1 })
  )

  let results = {
    getTotalRecords() {
      return totalRecords
    },
    async *[Symbol.asyncIterator]() {
      let next = run({ page, scrollId, skipCount: true })
      while (page <= Math.ceil(totalRecords / pageSize)) {
        let node = (results.node = await next)
        scrollId = node.context.scrollId
        next = run({ page: page + 1, scrollId, skipCount: true })
        page++
        // We return _source flattened onto the root results items because we're mostly
        // interested in the _source properties but may occasionally want other props like _id.
        // This will be removed with #28 when a contexture-elasticsearch upgrade is complete
        for (let r of _.map(
          flattenProp('_source'),
          resultField('results', node)
        ))
          yield r
      }
    },
  }
  return results
}
