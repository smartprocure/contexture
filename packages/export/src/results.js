import F from 'futil'
import _ from 'lodash/fp'
import { runWith, addIterator } from './utils'
import { flattenProp } from './futil'

// Will go away once results no longer wraps context in `response`
let resultField = (field, node) =>
  F.cascade([`response.${field}`, field], node.context)

export default async ({ service, tree, ...node}) => {
  let { page = 1, pageSize = 100 } = node
  let run = props =>
    runWith(
      service,
      tree,
      F.compactObject({ key: 'results', type: 'results', ...node, ...props })
    )
  let scrollId
  let totalRecords = resultField('totalRecords', await run({ pageSize: 1, page: 1 }))

  let result = {
    getTotalRecords() { return totalRecords },
    hasNext() { return page <= Math.ceil( totalRecords / pageSize) },
    async* [Symbol.asyncIterator]() {
      while (result.hasNext()) {
        let node = await run({ page, scrollId, skipCount: true })
        scrollId = node.context.scrollId
        page++
        // We return _source flattened onto the root result items because we're mostly
        // interested in the _source properties but may occasionally want other props like _id.
        // This will be removed with #28 when a contexture-elasticsearch upgrade is complete
        for (let r of _.map(flattenProp('_source'), resultField('results', node)))
          yield r
      }
    },
  }
  return result
}
