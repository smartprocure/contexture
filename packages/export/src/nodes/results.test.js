import _ from 'lodash/fp'
import results from './results'

describe('results', () => {
  let defaultTree = {
    key: 'root',
    children: [
      {
        key: 'terms',
        type: 'terms_stats',
      },
      {
        key: 'results',
        type: 'results',
      },
    ],
  }

  let simpleRecords = [
    { name: 'record1' },
    { name: 'record2' },
    { name: 'record3' },
  ]

  let getSimpleService = wrap =>
    jest.fn(tree => {
      let response = {
        totalRecords: 3,
        results: simpleRecords.map(_source => ({ _source })),
      }
      _.last(tree.children).context = wrap ? { response } : response
      return tree
    })

  let prepareSimpleStrategy = service => async (strategyParams = {}) => {
    let tree = _.cloneDeep(defaultTree)
    let include = ['a', 'b', 'c']
    let strategy = await results({
      service,
      tree,
      pageSize: 3,
      page: 1,
      totalPages: 1,
      include,
      sortField: 'a',
      sortDir: 'desc',
      ...strategyParams,
    })
    return strategy
  }
  let resultsTests = prepareSimpleStrategy => {
    it('retrieves the total records', async () => {
      let strategy = await prepareSimpleStrategy({ totalPages: 1 })
      expect(await strategy.getTotalRecords()).toBe(3)
    })
    it('retrieves records consistently with getNext', async () => {
      let strategy = await prepareSimpleStrategy({ page: 1 })
      let arr = []
      for await (const i of strategy) arr.push(i)
      expect(arr).toEqual(simpleRecords)
    })
  }
  describe(' with contexts wrapped in `response`', () => {
    resultsTests(prepareSimpleStrategy(getSimpleService(false)))
  })
  describe(' with contexts not wrapped in `response`', () => {
    resultsTests(prepareSimpleStrategy(getSimpleService(true)))
  })
})
