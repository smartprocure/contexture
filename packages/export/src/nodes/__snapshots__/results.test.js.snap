// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`results  with contexts not wrapped in \`response\` retrieves records 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "response": Object {
                "results": Array [
                  Object {
                    "_source": Object {
                      "name": "record1",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record2",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record3",
                    },
                  },
                ],
                "totalRecords": 3,
              },
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 1,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "response": Object {
                "results": Array [
                  Object {
                    "_source": Object {
                      "name": "record1",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record2",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record3",
                    },
                  },
                ],
                "totalRecords": 3,
              },
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "response": Object {
                "results": Array [
                  Object {
                    "_source": Object {
                      "name": "record1",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record2",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record3",
                    },
                  },
                ],
                "totalRecords": 3,
              },
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 2,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "response": Object {
                "results": Array [
                  Object {
                    "_source": Object {
                      "name": "record1",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record2",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record3",
                    },
                  },
                ],
                "totalRecords": 3,
              },
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 1,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "response": Object {
                "results": Array [
                  Object {
                    "_source": Object {
                      "name": "record1",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record2",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record3",
                    },
                  },
                ],
                "totalRecords": 3,
              },
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "response": Object {
                "results": Array [
                  Object {
                    "_source": Object {
                      "name": "record1",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record2",
                    },
                  },
                  Object {
                    "_source": Object {
                      "name": "record3",
                    },
                  },
                ],
                "totalRecords": 3,
              },
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 2,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
  ],
}
`;

exports[`results  with contexts wrapped in \`response\` retrieves records 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "results": Array [
                Object {
                  "_source": Object {
                    "name": "record1",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record2",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record3",
                  },
                },
              ],
              "totalRecords": 3,
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 1,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "results": Array [
                Object {
                  "_source": Object {
                    "name": "record1",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record2",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record3",
                  },
                },
              ],
              "totalRecords": 3,
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "results": Array [
                Object {
                  "_source": Object {
                    "name": "record1",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record2",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record3",
                  },
                },
              ],
              "totalRecords": 3,
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 2,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "results": Array [
                Object {
                  "_source": Object {
                    "name": "record1",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record2",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record3",
                  },
                },
              ],
              "totalRecords": 3,
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 1,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "results": Array [
                Object {
                  "_source": Object {
                    "name": "record1",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record2",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record3",
                  },
                },
              ],
              "totalRecords": 3,
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 1,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "results": Array [
                Object {
                  "_source": Object {
                    "name": "record1",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record2",
                  },
                },
                Object {
                  "_source": Object {
                    "name": "record3",
                  },
                },
              ],
              "totalRecords": 3,
            },
            "highlight": false,
            "include": Array [
              "a",
              "b",
              "c",
            ],
            "key": "results",
            "page": 2,
            "pageSize": 3,
            "skipCount": true,
            "sortDir": "desc",
            "sortField": "a",
            "totalPages": 1,
            "type": "results",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
  ],
}
`;
